### ISSUE TITLE
Кампании

### ISSUE THREAD

**Issue by biomancer** [fbkinfo/rosvybory#313]
Сейчас проект заточен под выборы Мэра Москвы, нужно обобщить и добавить работу с [кампаниями](https://github.com/fbkinfo/rosvybory/wiki/%D0%9F%D0%BE%D0%BB%D1%8C%D0%B7%D0%BE%D0%B2%D0%B0%D1%82%D0%B5%D0%BB%D1%8C%D1%81%D0%BA%D0%B8%D0%B5-%D1%81%D1%86%D0%B5%D0%BD%D0%B0%D1%80%D0%B8%D0%B8---%D0%90%D0%B4%D0%BC%D0%B8%D0%BD%D0%B8%D1%81%D1%82%D1%80%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5-%D0%BA%D0%B0%D0%BC%D0%BF%D0%B0%D0%BD%D0%B8%D0%B9).

**Comment by khodzha** [comment]
Как нужно указывать территорию проведения кампании?

**Comment by stiff** [comment]
Я думаю, что так как выборы проводятся какой-то Избирательной комиссией (ИК), поэтому правильнее всего доделать полноценное дерово *ИКов ( http://ru.wikipedia.org/wiki/Избирательная_комиссия ), и тогда просто нужно будет указать ИК для кампании.

**Comment by khodzha** [comment]
@stiff полноценное это всероссийское?

**Comment by stiff** [comment]
Не совсем понял вопрос, см. https://github.com/fbkinfo/rosvybory/blob/develop/app/models/uic.rb#L11 — сейчас только ТИК и УИК, а ЦИК'а или ИКС'ов нет. Базу конкретных комиссий сейчас наверно нет смысла искать, но чтобы была возможность залить при необходимости

**Comment by khodzha** [comment]
@stiff так что делать-то? добавить цик и икс? :)

**Comment by stiff** [comment]
К Uic'у — да :) можно переименовать его в Election Commission, чтобы сразу было понятно что это не только УИК.

Но основной запрос — про кампании, которых вообще нет. И если совсем делать "на все случаи жизни", то надо переписать ооочень много: например, определять роли пользователей в конкретной кампании, а не вообще. Не уверен что есть смысл это делать, потому что подозреваю, что проект будет разворачиваться на кампанию, они практически не пересекаются: территориальное деление за несколько лет может поменяться (расширение Москвы, объединение областей, новые города), люди другие, роли, всё другое одним словом.

**Comment by stiff** [comment]
Ещё вспомнил, административное деление сейчас только двух типов сделано (адм. регион и муниципальный регион), вот это то же что и с уиком — думаю было бы полезно добавить субъекты. Я начал переделывать, но не доделал, и сейчас получается User'ы и UserApp'ы и UserCurrentRole принципиально по-разному работают, что конечно же нужно унифицировать.

Было (и осталось в `User` и `UserApp`): adm_region есть всегда, а если есть region, то это мун. регион, иначе - нужно использовать adm_region, и куча if'ов.

Стало (`UserCurrentRole`): region — это фактический регион (любого типа) ~~, а в adm_region кешируется (чтобы избежать тяжёлый бой с activeadmin'ом, не помню что именно, я после нескольких экспериментов сдался) административный регион выбранного региона, если он есть~~ **upd:** это я обманул, оказывается всё-таки кеш я убрал, но не очень расширяемо, из-за того как эти поля задаются в интерфейсе: https://github.com/fbkinfo/rosvybory/blob/develop/app/models/concerns/adm_region_delegation.rb

**Comment by khodzha** [comment]
@stiff я так и не понял до конца, что делать нужно. начну пилить избирательные комиссии, как появится что-то относительно правдоподобное - позову ревьюить

**Comment by stiff** [comment]
@khodzha да, так самое правильное — если что-то никому не понятно, значит оно пока не нужно :)
